gdal/port/cpl_userfaultfd.cpp

```diff
@@ -132,44 +149,144 @@ static void * fault_handler(void * ptr)
+      		 "fault_handler: sigaction(SIGSEGV) failed");
+      	break;
+      }
+      if (sigaction(SIGBUS, &bus, &old_bus) == -1) {
+      	CPLError(CE_Failure, CPLE_AppDefined,
+      		 "fault_handler: sigaction(SIGBUS) failed");
+      	break;
       }
-    }
 
-    // Copy data into page
-    if (VSIFSeekL(file, offset, SEEK_SET)) break;
-    if (VSIFReadL(ctx->page_ptr, bytes_needed, 1, file) != 1) break;
-    pages_used++;
-
-    uffdio_copy.src = reinterpret_cast<uintptr_t>(ctx->page_ptr);
-    uffdio_copy.dst = fault_addr;
-    uffdio_copy.len = static_cast<uintptr_t>(ctx->page_size);
-    uffdio_copy.mode = 0;
-    uffdio_copy.copy = 0;
-    if (ioctl(ctx->uffd, UFFDIO_COPY, &uffdio_copy) == -1) break;
+      // WARNING LACK OF THREAD-SAFETY.
+      //
+      // For example, if a user program (or another part of the
+      // library) installs a SIGSEGV or SIGBUS handler from another
+      // thread after this one has installed its handlers but before
+      // this one uninstalls its handlers, the intervening handler
+      // will be eliminated.  There are other examples, as well, but
+      // there can only be a problems with other threads because the
+      // faulting thread is blocked here.
+      //
+      // This implies that one should not create `/vsimem` handles
+      // while other threads are actively generating faults that use
+      // this mechanism.
+      //
+      // Having multiple active threads that use this mechanism but
+      // with no changes to signal-handling in other threads is NOT a
+      // problem.
+
+      // Step 2
+      if (mprotect(ctx->vma_ptr, ctx->vma_size, PROT_NONE) == -1) {
+      	CPLError(CE_Failure, CPLE_AppDefined,
+      		 "fault_handler: mprotect() failed");
+      	break;
+      }
+
+      // Step 3
+      if (ioctl(ctx->uffd, UFFDIO_UNREGISTER, &ctx->uffdio_register)) {
+      	CPLError(CE_Failure, CPLE_AppDefined,
+      		 "fault_handler: ioctl(UFFDIO_UNREGISTER) failed");
+      	break;
+      }
+      ctx->vma_ptr = mmap(ctx->vma_ptr, ctx->vma_size, PROT_NONE, MAP_PRIVATE|MAP_ANONYMOUS|MAP_FIXED, -1, 0);
+      if (ctx->vma_ptr == BAD_MMAP) {
+      	CPLError(CE_Failure, CPLE_AppDefined,
+      		 "fault_handler: mmap() failed");
+      	break;
+      }
+      ctx->pages_used = 0;
+      if (ioctl(ctx->uffd, UFFDIO_REGISTER, &ctx->uffdio_register)) {
+      	CPLError(CE_Failure, CPLE_AppDefined,
+      		 "fault_handler: ioctl(UFFDIO_REGISTER) failed");
+      	break;
+      }
+
+      // Step 4.  Problem: A thread might attempt to read here (before
+      // the mprotect) and recieve a SIGSEGV or SIGBUS.
+      if (mprotect(ctx->vma_ptr, ctx->vma_size, PROT_READ) == -1) {
+      	CPLError(CE_Failure, CPLE_AppDefined,
+      		 "fault_handler: mprotect() failed");
+      	break;
+      }
+
+      // Step 5.  Solution: Cannot unregister special handlers before
+      // any such threads have been handled by them.
+      usleep(10000);
```

Comment:
I would like to find better way to do this.

The issue appears to be that you might have a thread that generates a signal before `mprotect` above executes, gets interrupted, then does not resume until after the handlers have been removed.  1/100 of a second seems like enough time for signaled threads to get into the special handler before it is removed.